File: src/workflow/base/agent_db.ts
import { Handler } from './handler';

export class AgentDB {
    // 基本信息字段（数据库中必须的字段）
    public cid: string = "";
    public apiv: string = "";
    public apisys: string = "";
    public apiobj: string = "";
    public agentname: string = "";
    public description: string = "";
    public maxcopy: number = 1;      // 最大并发数
    public version: string = "1.0.0";
    public errsec: number = 86400;   // 多少秒没成功才算失败

    // 状态和时间字段
    public state: string = "active";  // 状态：active(启用), inactive(禁用)
    public lastheartbeat: Date | null = null;  // 最后心跳时间，用于监控代理存活状态
    public currentcopy: number = 0;      // 当前并发数

    // 时间和统计字段
    public lastoktime: Date | null = null;      // 最后成功时间
    public lastokinfo: string = "";      // 成功信息
    public starttime: Date | null = null;
    public endtime: Date | null = null;
    public lasterrinfo: string = "";    // 错误信息

    // 价格成本相关字段
    public pricebase: number = 1.0;  // 基础价格
    public price: number = 1.0;  // 当前价格
    public costunit: number = 0.0;  // 单位成本
    public profittarget: number = 0.2;  // 目标利润率
    public profittotal: number = 0.0;  // 总利润
    public costtotal: number = 0.0;  // 总成本
    public revenuetotal: number = 0.0;  // 总收入
    public roi: number = 0.0;  // 投资回报率

    // 执行统计相关字段
    public successcount: number = 0;  // 成功执行次数
    public runcount: number = 0;  // 总执行次数
    public successrate: number = 0.0;  // 成功率
    public errorcount: number = 0;  // 错误次数

    // 描述信息
    public costdescription: string = "";  // 成本描述
    public pricedescription: string = "";  // 价格描述

    // 系统字段
    public upby: string = "";
    public uptime: Date | null = null;  // 由数据库自动更新
    public idpk: number | null = null;  // 由数据库自动生成
    public id: string | null = null;  // 唯一标识符

    // 备注字段
    public remark: string = "";
    public remark2: string = "";
    public remark3: string = "";
    public remark4: string = "";
    public remark5: string = "";
    public remark6: string = "";

    // 动态字段
    public jsobj: Record<string, any> = {};
    public inputdata: Record<string, any> = {};
    public outputdata: Record<string, any> = {};

    // 处理器映射（内存中缓存，用于快速访问）
    private handlersByType: Map<string, Map<string, any>> = new Map(); // 按类型和能力分类: type -> capability -> handler
    private handlersByName: Map<string, any> = new Map(); // 按处理器名称分类: handler_name -> handler

    constructor(json_data?: Record<string, any>) {
        if (json_data) {
            this.setFromJson(json_data);
        }
    }

    public setFromJson(json_data: Record<string, any>): this {
        for (const key in json_data) {
            if (Object.prototype.hasOwnProperty.call(json_data, key)) {
                // 特殊处理 handlers 字段，将其转换为处理器映射
                if (key === 'handlers' && Array.isArray(json_data[key])) {
                    this.loadHandlersFromJsonData(json_data[key]);
                } else {
                    // 处理日期类型字段
                    if (key.includes('time') || key.includes('date')) {
                        if (json_data[key]) {
                            (this as any)[key] = new Date(json_data[key]);
                        }
                    } else {
                        (this as any)[key] = json_data[key];
                    }
                }
            }
        }

        // 如果jsobj被更新，也更新inputdata和outputdata
        if (json_data.jsobj && typeof json_data.jsobj === 'object') {
            this.jsobj = json_data.jsobj;
            this.inputdata = this.jsobj.inputdata || {};
            this.outputdata = this.jsobj.outputdata || {};
        }

        return this;
    }

    public toJson(): Record<string, any> {
        const result: Record<string, any> = {};

        for (const key in this) {
            if (Object.prototype.hasOwnProperty.call(this, key)) {
                // 跳过一些动态计算或不应被序列化的属性
                if (key !== 'inputdata' && key !== 'outputdata' && key !== 'handlersByType' && key !== 'handlersByName') {
                    const value = (this as any)[key];
                    // 处理日期类型字段
                    result[key] = value instanceof Date ? value.toISOString() : value;
                }
            }
        }

        // 将处理器转换为JSON数组格式以便保存到数据库
        result.handlers = this.convertHandlersToJsonArray();

        // 将inputdata和outputdata合并到jsobj中
        this.jsobj.inputdata = this.inputdata;
        this.jsobj.outputdata = this.outputdata;
        result.jsobj = this.jsobj;

        return result;
    }

    // 从JSON数据加载handlers到内存映射
    private loadHandlersFromJsonData(handlersJson: any[]): void {
        // 清空现有的处理器
        this.handlersByType.clear();
        this.handlersByName.clear();

        // 遍历JSON数组，将每个handler转换为Handler实例
        for (const handlerData of handlersJson) {
            const handler = new Handler();
            handler.setFromJson(handlerData);  // 使用Handler类的setFromJson方法

            // 使用handler的type和capability作为映射的键
            const type = handler.apisys || 'default';
            const capability = handler.capability;
            const handlerName = handler.handler;

            // 添加到按类型和能力分类的映射中
            if (!this.handlersByType.has(type)) {
                this.handlersByType.set(type, new Map());
            }
            this.handlersByType.get(type)?.set(capability, handler);

            // 添加到按处理器名称分类的映射中
            if (handlerName) {
                this.handlersByName.set(handlerName, handler);
            }
        }
    }

    // 将handlers转换为JSON数组格式，用于保存到数据库
    private convertHandlersToJsonArray(): any[] {
        const result: any[] = [];

        // 从handlersByType映射中收集所有处理器
        for (const [, capabilities] of this.handlersByType) {
            for (const [, handler] of capabilities) {
                // 使用Handler类的toJson方法
                result.push((handler as Handler).toJson());
            }
        }

        // 确保handlersByName中的处理器也包含在结果中（避免重复）
        for (const [, handler] of this.handlersByName) {
            const handlerJson = (handler as Handler).toJson();
            // 检查是否已存在于结果中（通过id或handler名称）
            const exists = result.some(h =>
                h.id === handlerJson.id || h.handler === handlerJson.handler
            );
            if (!exists) {
                result.push(handlerJson);
            }
        }

        return result;
    }

    // 检查代理是否处于活动状态
    public isActive(): boolean {
        return this.state === "active";
    }

    // 添加处理器到内存映射
    public addHandler(handler: any): void {
        const { type, capability, handler: handlerName } = handler;

        // 添加到按类型和能力分类的映射中
        if (!this.handlersByType.has(type)) {
            this.handlersByType.set(type, new Map());
        }
        this.handlersByType.get(type)?.set(capability, handler);

        // 添加到按处理器名称分类的映射中
        if (handlerName) {
            this.handlersByName.set(handlerName, handler);
        }
    }

    // 获取处理器
    // 按类型和能力获取处理器
    public getHandlerByCapability(type: string, capability: string): any | null {
        return this.handlersByType.get(type)?.get(capability) || null;
    }

    // 按处理器名称获取处理器
    public getHandlerByName(handlerName: string): any | null {
        return this.handlersByName.get(handlerName) || null;
    }

    // 获取所有处理器（按类型和能力）
    public getAllHandlersByType(): Map<string, Map<string, any>> {
        return this.handlersByType;
    }

    // 获取所有处理器（按名称）
    public getAllHandlersByName(): Map<string, any> {
        return this.handlersByName;
    }

    // 获取指定类型的所有处理器
    public getHandlersByType(type: string): Map<string, any> | null {
        return this.handlersByType.get(type) || null;
    }


    // 执行处理器（按处理器名称）
    public async executeHandler(handlerName: string, params: Record<string, any>): Promise<Record<string, any>> {
        const handler = this.getHandlerByName(handlerName);
        if (!handler) {
            throw new Error(`Handler not found for name: ${handlerName}`);
        }

        // 执行处理器 - 所有处理器都应该是Handler实例
        return await handler.execute(params);
    }


}

File: src/workflow/base/agent_db.py
"""
AgentDB类 - 用于映射workflow_agent表的字段
"""


class AgentDB:
    """
    Agent数据库模型类，映射workflow_agent表的字段
    """
    
    def __init__(self, json_data=None):
        """
        初始化AgentDB实例
        严格按照workflow_agent表结构映射
        
        :param json_data: 可选的JSON字典数据，用于初始化实例
        """
        # 基本信息字段（数据库中必须的字段）
        self.cid = ""
        self.apiv = ""
        self.apisys = ""
        self.apiobj = ""
        self.agentname = ""
        self.description = ""
        self.maxcopy = 1      # 最大并发数
        self.version = "1.0.0"
        self.errsec = 86400   # 多少秒没成功才算失败
        
        # 状态和时间字段
        self.state = "active"  # 状态：active(启用), inactive(禁用)
        self.lastheartbeat = None  # 最后心跳时间，用于监控代理存活状态
        self.currentcopy = 0      # 当前并发数
        
        # 时间和统计字段
        self.lastoktime = None      # 最后成功时间
        self.lastokinfo = ""      # 成功信息
        self.starttime = None
        self.endtime = None
        self.lasterrinfo = ""    # 错误信息
        
        # 价格成本相关字段
        self.pricebase = 1.0  # 基础价格
        self.price = 1.0  # 当前价格
        self.costunit = 0.0  # 单位成本
        self.profittarget = 0.2  # 目标利润率
        self.profittotal = 0.0  # 总利润
        self.costtotal = 0.0  # 总成本
        self.revenuetotal = 0.0  # 总收入
        self.roi = 0.0  # 投资回报率
        
        # 执行统计相关字段
        self.successcount = 0  # 成功执行次数
        self.runcount = 0  # 总执行次数
        self.successrate = 0.0  # 成功率
        self.errorcount = 0  # 错误次数
        
        # 描述信息
        self.costdescription = ""  # 成本描述
        self.pricedescription = ""  # 价格描述
        
        # 系统字段
        self.upby = ""
        self.uptime = None  # 由数据库自动更新
        self.idpk = None  # 由数据库自动生成
        self.id = None  # 唯一标识符
        
        # 备注字段
        self.remark = ""
        self.remark2 = ""
        self.remark3 = ""
        self.remark4 = ""
        self.remark5 = ""
        self.remark6 = ""
        
        # 运行时动态字段（不存储在数据库中）
        self.handlers = []  # 存储此Agent支持的handlers
        
        # 检查jsobj字段是否存在，如果不存在则初始化
        if not hasattr(self, 'jsobj'):
            self.jsobj = {}
        if not hasattr(self, 'inputdata'):
            self.inputdata = {}
        if not hasattr(self, 'outputdata'):
            self.outputdata = {}
        
        # 如果提供了JSON数据，则使用它初始化实例
        if json_data is not None:
            self.set_from_json(json_data)

    def set_from_json(self, json_data):
        """
        从JSON数据设置属性
        """
        for key, value in json_data.items():
            if hasattr(self, key):
                setattr(self, key, value)
        
        # 如果jsobj被更新，也更新inputdata和outputdata
        if 'jsobj' in json_data and isinstance(json_data['jsobj'], dict):
            self.jsobj = json_data['jsobj']
            self.inputdata = self.jsobj.get('inputdata', {})
            self.outputdata = self.jsobj.get('outputdata', {})
        
        return self
        
    def get_from_json(self):
        """
        将对象转换为JSON格式
        确保inputdata和outputdata正确保存到jsobj中
        """
        # 创建当前对象的属性副本
        result = {}
        
        # 获取所有属性
        for key, value in self.__dict__.items():
            # 跳过一些动态计算的属性
            if key not in ['inputdata', 'outputdata']:
                result[key] = value
        
        # 将inputdata和outputdata合并到jsobj中
        self.jsobj['inputdata'] = self.inputdata
        self.jsobj['outputdata'] = self.outputdata
        result['jsobj'] = self.jsobj
        
        return result
    
    def add_handler(self, handler):
        """
        添加一个handler到此Agent
        :param handler: Handler对象
        """
        from .handler import Handler
        if isinstance(handler, Handler):
            # 确保handler的idagent与此agent的id匹配
            handler.idagent = self.id or self.agentname
            self.handlers.append(handler)
        else:
            raise TypeError("handler必须是Handler类的实例")
    
    def get_handlers_by_capability(self, capability):
        """
        根据capability获取handlers
        :param capability: 功能类型
        :return: 匹配的handlers列表
        """
        return [h for h in self.handlers if h.capability == capability and h.state == 'active']
    
    def can_handle(self, capability):
        """
        检查Agent是否能处理特定capability
        :param capability: 功能类型
        :return: 是否能处理
        """
        return len(self.get_handlers_by_capability(capability)) > 0